### ISSUE TITLE
The libdftatom.so is linked against systemwide libgfortran.so instead of the conda-forge

### ISSUE THREAD

**Issue by certik** [conda-forge/dftatom-feedstock#1]
Issue:

The `$CONDA_PREFIX/lib/libdftatom.so` shared library is built against the system Fortran runtime, not Conda's runtime:
```
$ ldd $CONDA_PREFIX/lib/libdftatom.so
	linux-vdso.so.1 (0x00007fff249e9000)
	libgfortran.so.4 => /usr/lib/x86_64-linux-gnu/libgfortran.so.4 (0x000014bc4385e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000014bc434c0000)
	libgcc_s.so.1 => /home/certik/ext/miniconda3/envs/test15/lib/./libgcc_s.so.1 (0x000014bc43dd3000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x000014bc432a1000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000014bc42eb0000)
	libquadmath.so.0 => /home/certik/ext/miniconda3/envs/test15/lib/./libquadmath.so.0 (0x000014bc43d99000)
	/lib64/ld-linux-x86-64.so.2 (0x000014bc43c3d000)
```
As you can see, the gcc runtime is correctly linked from Conda, but `libgfortran.so` is not.

This is caused by https://github.com/conda-forge/gfortran_osx-64-feedstock/issues/3

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /home/certik/ext/miniconda3/envs/test15:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
dftatom                   1.0.4                h4333106_0    conda-forge
libgcc-ng                 9.1.0                hdf63c60_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : test15
    active env location : /home/certik/ext/miniconda3/envs/test15
            shell level : 1
       user config file : /home/certik/.condarc
 populated config files : /home/certik/.condarc
          conda version : 4.7.5
    conda-build version : not installed
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /home/certik/ext/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/certik/ext/miniconda3/pkgs
                          /home/certik/.conda/pkgs
       envs directories : /home/certik/ext/miniconda3/envs
                          /home/certik/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.5 requests/2.22.0 CPython/3.7.3 Linux/4.15.0-54-generic ubuntu/18.04.2 glibc/2.27
                UID:GID : 15148:15148
             netrc file : None
           offline mode : False
```
</details>

**Comment by certik** [comment]
@wolfv I discovered this issue. It's caused by a [bug](https://github.com/conda-forge/gfortran_osx-64-feedstock/issues/3) I reported some time ago, but I don't know how to fix it.

**Comment by beckermr** [comment]
This is incorrect. You are building with gfortran 4.8.5 which was built by anaconda inc and then copied into the conda-forge channel. It is not the bug you are reporting above.

Here is the exact line from the build.

> gfortran_osx-64: 4.8.5-h22b1bf0_8           conda-forge

Note also that the bug you reported is standard, expected behavior given how the compilers are shipped and conda in general.

**Comment by isuruf** [comment]
It looks like on linux, the dependency `libgfortran-ng` is not added. Definitely a conda-build bug

**Comment by certik** [comment]
@beckermr thanks for the feedback. Can you be more specific what you mean by "You are building with gfortran 4.8.5"? The `dftatom` package is built by conda-forge on the CI with (what I assume) are conda-forge compilers.

@isuruf when I do `conda install -c conda-forge libgfortran-ng`, it still links against the system-wide runtime.

**Comment by isuruf** [comment]
@certik, this is an issue when the package is being built. If `libgfortran-ng` is not in host when the package is being built, then manually installing it wouldn't help

**Comment by certik** [comment]
@isuruf I see. How do we fix it? By adding it as a dependency explicitly into `dftatom`?

**Comment by isuruf** [comment]
You could do that, but `conda-build` should do it automatically. Not sure why

**Comment by beckermr** [comment]
There are two versions of gfortran in the conda-forge channels. See the packages here: https://anaconda.org/conda-forge/gfortran_osx-64/files

The version 4.8.5 package was added by @isuruf manually and I assume matches that from defaults.
The version 7.3.0 package was built in the feedstock you linked and contributed by myself. 

Conda-forge is globally pinned at version 4.8.5 of gfortran at the moment. See the global conda-forge pinning file. 

Thus the problem you are finding is not an issue with the gfortran 7.3.0 build. In fact, if you check the build logs on azure for osx, you will see that the version 7.3.0 package is not even used.

**Comment by isuruf** [comment]
@beckermr, we use 4.8.5 of gfortran only for OSX. gfortran 7.3.0 is used in Linux builds. This issue is about a linux build, not OSX.

**Comment by beckermr** [comment]
Ahh crap. I missed this was linux. :( Ignore everything I said.

**Comment by beckermr** [comment]
Given that the linked issue is on osx build, this is definitely not an issue in that feedstock!

**Comment by jjhelmus** [comment]
Most like a result of conda/conda-build#3610.

**Comment by certik** [comment]
@beckermr Thanks for your help and my apologies I didn't mean to insinuate that there is anything wrong with your package or your work, I simply thought the underlying cause is the same (that gfortran doesn't automatically set rpath to the correct runtime library), both on macOS and Linux. Btw, here is how GCC has to be patched to produce the correct rpath:

https://github.com/hashdist/hashstack/blob/cc69c4b8ddbf82fe2410c4188f5d6ad2980c768b/pkgs/gcc/rpath.patch

We have used this in Hashdist, and it works great.

I don't fully understand the Conda requirements and constraints, so I don't know if this patch would be viable. I also don't understand how the rpath could be set without this patch: does Conda somehow patch the binaries?

**Comment by isuruf** [comment]
> I also don't understand how the rpath could be set without this patch: does Conda somehow patch the binaries?

Yes, conda patches the binaries to fix the rpath.

**Comment by certik** [comment]
@isuruf I know Conda patches the binaries for hardcoded paths that are within the build directory, however the Fortran's runtime library typically does not link into the build directory, and so that's why I am surprised that Conda can figure this out.

**Comment by isuruf** [comment]
Ah, we have `-L$PREFIX/lib, -Wl,-rpath,$PREFIX/lib` in `LDFLAGS`, so `libgfortran.so` is found.

**Comment by beckermr** [comment]
no worries @certik! My apologies for coming off as prickly. :)

**Comment by certik** [comment]
@isuruf I see. And on user's computer, the $PREFIX gets rewritten, so it still finds the correct `libgfortran.so`. Ok, that should work, once the conda-build issue is fixed.
